File: leetcode/python/Binary Tree Preorder Traversal/Binary_Tree_Preorder_Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from TreeNode import TreeNode
class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        if root:
            return []
        res = []
        return self._preorderTraversal(root,res)
    def _preorderTraversal(self,root,res):
        if root is None:
            return res
        res.append(root.val)
        self._preorderTraversal(root.left,res)
        self._preorderTraversal(root.right,res)
        return res

solu = Solution()
root = TreeNode(1)
solu.preorderTraversal(root)

File: leetcode/cpp/BinaryTreePreorderTraversal.cpp
//
// Created by zhangqi on 2017/11/15.
// Time:  O(n)
// Space: O(n)

#include <vector>
#include <stack>

#include "TreeNode.h"

using namespace std;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<pair<TreeNode*, bool>> s;
        s.push(make_pair(root, false));
        while(!s.empty()){
            bool visit;
            TreeNode *curr;
            tie(curr, visit) = s.top();
            s.pop();
            if(curr == nullptr){
                continue;
            }

            if(visit){
                ans.push_back(curr->val);
            } else{
                s.push(make_pair(curr->right,false));
                s.push(make_pair(curr->left, false));
                s.push(make_pair(curr,true));
            }
        }
        return ans;
    }
};